### DISCUSSION TITLE
"Reload libraries" related

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#79] [Q&A]
Hi,
I have a peculiar situation:

1. I open KLayout and select my custom chip from the library to add it to the view. Works fine and as expected.
2. I simply hit reload libraries from KQC menu and I get an error which is something like this:

![image](https://github.com/iqm-finland/KQCircuits/assets/97676306/03d2f545-fb08-4b73-81fd-0833fb8eef93)

Have you seen a similar situation before? It always works the first time I open Klayout, but fails everytime if I simply reload libraries.

**Comment by qpavsmi** [comment]
Hi! We have internally noticed that "Reload libraries" sometimes fails for our designs too, but haven't had the bandwidth to fix it yet.

Is it that the available KQC elements work fine with "Reload libraries" but only your custom chip fails? It could help us if you could share for example some stripped down version of your code that breaks when "Reload libraries" is performed.

In any case, we have it in our backlog to fix the command for our designs, and once we do that I am quite sure it would fix it for you too. But for the time being, the safest way to reload changes is unfortunately to restart KLayout.

**Comment by Qvadi** [comment]
Hi qpavsmi,
I have a custom chip with a similar implementation of a custom rf tsv using the existing flip_chip connector module. This is how it looks: tsv (top), bump (bottom)
![image](https://github.com/iqm-finland/KQCircuits/assets/97676306/88f69115-fe3d-46a8-8ac6-7e46e54bc470)

I have added this tsv element as a layer changing element in defaults.py. It works the first time when this tsv is part of the waveguide_composite definition. But when I reload libraries, it fails. As an example check: `Node(pya.DPoint(x,y), face_id="2b1")` which uses the bump as a default face changing element, this works all the time even if you reload libraries. But if you do something like this: `Node(pya.DPoint(x,y), FlipChipConnectorRf, face_id="2b1")` works the first time, but fails after reloading libraries in spite `FlipChipConnectorRf` being the default node changing element. The same happens with the tsv.

**Reply by Qvadi** [reply]
You can see from the image, that both tsv and bump has similar port names.

**Reply by Qvadi** [reply]
Here in the waveguide_composite element code, 
![image](https://github.com/iqm-finland/KQCircuits/assets/97676306/b64d96d5-28e7-476a-a9b3-d04f044b3d0e)
the error that I initially shared, happens at line 503 as well. The code is not considering these elements as face_changing even though they are defined that way. I am not sure, why the code skips this part
